docs/xdsl-introduction.ipynb

```diff
@@ -0,0 +1,845 @@
+      "  scf.yield(%6 : !i1)\n",
+      "}\n"
+     ]
+    }
+   ],
+   "source": [
+    "walker = PatternRewriteWalker(\n",
+    "    GreedyRewritePatternApplier([\n",
+    "        ConstantFolding()\n",
+    "    ]),\n",
+    "    walk_regions_first=True,\n",
+    "    apply_recursively=True,\n",
+    "    walk_reverse=False)\n",
+    "walker.rewrite_module(filtered)\n",
+    "printer.print_op(filtered)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2508e53b-963d-4ea9-87bb-ba3b150fbade",
+   "metadata": {},
+   "source": [
+    "the above introduced some variables which are no longer used (e.g. `%3` and `%4`), therefore can be safely removed from the code."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "2f9b0e45-a93f-4060-a36b-1e0a176e6ea6",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "%1 : !sql.bag = sql.filter(%0 : !sql.bag) {\n",
+      "^0(%9 : !i32):\n",
+      "  %8 : !i32 = arith.constant() [\"value\" = 10 : !i32]\n",
+      "  %6 : !i1 = arith.cmpi(%9 : !i32, %8 : !i32) [\"predicate\" = 4 : !i64]\n",
+      "  scf.yield(%6 : !i1)\n",
+      "}\n"
+     ]
+    }
+   ],
+   "source": [
+    "@dataclass\n",
+    "class DeadConstantElim(RewritePattern):\n",
+    "\n",
+    "  @op_type_rewrite_pattern\n",
+    "  def match_and_rewrite(self, op: arith.Constant, rewriter: PatternRewriter):\n",
+    "        if len(op.result.uses) == 0:\n",
+    "            rewriter.erase_matched_op()\n",
+    "\n",
+    "walker = PatternRewriteWalker(\n",
+    "    GreedyRewritePatternApplier([\n",
+    "        DeadConstantElim()\n",
+    "    ]),\n",
+    "    walk_regions_first=True,\n",
+    "    apply_recursively=True,\n",
+    "    walk_reverse=False)\n",
+    "walker.rewrite_module(filtered)\n",
+    "printer.print_op(filtered)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5bc69aa0-c647-4f7c-87c2-3d6a6e2551b3",
+   "metadata": {},
+   "source": [
+    "And now, the transformed filter operation is equivalent to:\n",
+    "```\n",
+    "filter(%table, (lambda t: t.a > 10))\n",
+    "```"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "97da7b96-b641-4915-b6b1-b38ca4cf00c5",
+   "metadata": {},
+   "source": [
+    "### Summary\n",
+    "So, we have now defined a mini-dialect using many of xDSL's features and concepts introduced earlier, including:\n",
+    "- Operations: we modelled the SQL dialect using `select` and `filter` operations\n",
+    "- Attributes: we used attributes to encode compile-time available information, from types (the `Bag` type), to constant expressions (e.g. the \"cmpi\" attribute within `arith.Cmpi`)\n",
+    "- Mixing dialects: we utilised features from other dialects (`arith`, `scf`, `builtin`)\n",
+    "- Regions and blocks: we used the operation's _regions_ to model the lambda within `sql.filter` since regions correspond to an SCF-like representation of code.\n",
```

Comment:
```suggestion
    "- Regions and blocks: we used the operation's _regions_ to model the lambda within `sql.filter` since regions allow nesting of code.\n",
```